integration-cli/docker_api_containers_test.go

```diff
@@ -1947,3 +1947,46 @@ func (s *DockerSuite) TestContainerKillCustomStopSignal(c *check.C) {
 	err = waitInspect(id, "{{.State.Running}} {{.State.Restarting}}", "false false", 30*time.Second)
 	c.Assert(err, checker.IsNil)
 }
+
+// Regression test for #34307
+// Make sure that a container shares its namespace with ID.
+// Test renaming a container which namespace is shared by another container.
+func (s *DockerSuite) TestContainerShareNamespaceByName(c *check.C) {
+	// Namespace is not supported on Windows
+	testRequires(c, DaemonIsLinux, SameHostDaemon, NotUserNamespace)
+
+	name := "base"
+	out := runSleepingContainer(c, "--name", name)
+	id := strings.TrimSpace(out)
+	expected := "container:" + id
+
+	dockerCmd(c, "run", "--name=share", "--ipc", "container:"+name, "--pid", "container:"+name, "--net", "container:"+name, "busybox")
```

Comment:
@dnephin I have modified `adaptSharedNamespaceContainer()` to accept a `containerGetter` interface as you suggested, here is the demo:
```go
diff --git a/daemon/daemon_unix.go b/daemon/daemon_unix.go
index b3eaa2f68..8bed31da3 100644
--- a/daemon/daemon_unix.go
+++ b/daemon/daemon_unix.go
@@ -64,6 +64,10 @@ const (
        cgroupSystemdDriver = "systemd"
 )

+type containerGetter interface {
+       GetContainer(string) (*container.Container, error)
+}
+
 func getMemoryResources(config containertypes.Resources) *specs.LinuxMemory {
        memory := specs.LinuxMemory{}

@@ -285,7 +289,7 @@ func (daemon *Daemon) adaptContainerSettings(hostConfig *containertypes.HostConf
                hostConfig.IpcMode = containertypes.IpcMode(m)
        }

-       daemon.adaptSharedNamespaceContainer(hostConfig)
+       adaptSharedNamespaceContainer(daemon, hostConfig)

        var err error
        opts, err := daemon.generateSecurityOpt(hostConfig)
@@ -304,7 +308,7 @@ func (daemon *Daemon) adaptContainerSettings(hostConfig *containertypes.HostConf
 // adaptSharedNamespaceContainer replaces a container name with its container ID when
 // a container shares its namespace by container name,so as to keep the namespace
 // sharing connection when renaming the container which namespaces be shared.
-func (daemon *Daemon) adaptSharedNamespaceContainer(hostConfig *containertypes.HostConfig) error {
+func adaptSharedNamespaceContainer(daemon containerGetter, hostConfig *containertypes.HostConfig) error {
        containerPrefix := "container:"
        if hostConfig.PidMode.IsContainer() {
                pidContainer := hostConfig.PidMode.Container()
```

> That way you can easily provide a fake which returns the values you care about

But what do you mean by `a fake`? If I understood correctly, you meant I should define a `fakedaemon`  and implement `GetContainer()` ( I think it's complicated to implement this method ), then use that `fakedaemon` to test func `adaptSharedNamespaceContainer`? Could you please give me more infomation?

I noticed a unittest named `TestAdjustCPUShares`, I think I can write a similar one without any change of `adaptSharedNamespaceContainer()` if it's acceptable.